src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.cs

```diff
@@ -443,30 +430,46 @@ private static int IndexOf(this JsonObject jsonObject, string key)
     }
 #endif
 
-    private static TAttribute? ResolveAttribute<TAttribute>(this JsonSchemaExporterContext ctx)
-        where TAttribute : Attribute
-    {
-        // Resolve attributes from locations in the following order:
-        // 1. Property-level attributes
-        // 2. Parameter-level attributes and
-        // 3. Type-level attributes.
-        return
-#if NET9_0_OR_GREATER
-            GetAttrs(ctx.PropertyInfo?.AttributeProvider) ??
-            GetAttrs(ctx.PropertyInfo?.AssociatedParameter?.AttributeProvider) ??
-#else
-            GetAttrs(ctx.PropertyAttributeProvider) ??
-            GetAttrs(ctx.ParameterInfo) ??
-#endif
-            GetAttrs(ctx.TypeInfo.Type);
-
-        static TAttribute? GetAttrs(ICustomAttributeProvider? provider) =>
-            (TAttribute?)provider?.GetCustomAttributes(typeof(TAttribute), inherit: false).FirstOrDefault();
-    }
-
     private static JsonElement ParseJsonElement(ReadOnlySpan<byte> utf8Json)
     {
         Utf8JsonReader reader = new(utf8Json);
         return JsonElement.ParseValue(ref reader);
     }
+
+    /// <summary>The equatable key used to look up cached schemas.</summary>
+    private readonly record struct SchemaGenerationKey
+    {
+        public SchemaGenerationKey(
+            Type? type,
+            string? parameterName,
+            string? description,
+            bool hasDefaultValue,
+            object? defaultValue,
+            AIJsonSchemaCreateOptions options)
+        {
+            Type = type;
+            ParameterName = parameterName;
+            Description = description;
+            HasDefaultValue = hasDefaultValue;
+            DefaultValue = defaultValue;
+            IncludeSchemaKeyword = options.IncludeSchemaKeyword;
+            DisallowAdditionalProperties = options.DisallowAdditionalProperties;
+            IncludeTypeInEnumSchemas = options.IncludeTypeInEnumSchemas;
+            RequireAllProperties = options.RequireAllProperties;
+            FilterDisallowedKeywords = options.FilterDisallowedKeywords;
+            TransformSchemaNode = options.TransformSchemaNode;
+        }
+
+        public Type? Type { get; }
+        public string? ParameterName { get; }
+        public string? Description { get; }
+        public bool HasDefaultValue { get; }
+        public object? DefaultValue { get; }
+        public bool IncludeSchemaKeyword { get; }
+        public bool DisallowAdditionalProperties { get; }
+        public bool IncludeTypeInEnumSchemas { get; }
+        public bool RequireAllProperties { get; }
+        public bool FilterDisallowedKeywords { get; }
+        public Func<AIJsonSchemaCreateContext, JsonNode, JsonNode>? TransformSchemaNode { get; }
```

Comment:
Note that this is the first time a type with reference equality is being introduced to the components of the key. While there is precedent in us doing this elsewhere (notably in the `JsonSerializerOptions` cache in STJ), an application using ephemeral transformer delegates would quickly saturate the cache.

I briefly considered nesting another ConditionalWeakTable that is keyed on the delegate, but I suspect this is overkill.

Reply:
A viable alternative would involve adding the JSO as a property of `AIJsonSchemaCreateOptions` and then key the existing CWT on that. It is a breaking change though.

Reply:
One possible option is inserting the JSO as a property on the `AIJsonCreateSchemaOptions` and keying the CWT on instances of the latter. This would allow caching of schemas for standalone types but disable caching of AIFunction parameters since they provide their own JSO -- this is probably fine though since schemas are created and stored when the AIFunction is constructed.

Reply:
I ended up disabling caching for inputs that specify a transformer -- https://github.com/dotnet/extensions/pull/5677/commits/bffe54d0aa9eaf96fc0f316d18f7411dd0e5c413

Reply:
> I ended up disabling caching for inputs that specify a transformer -- [bffe54d](https://github.com/dotnet/extensions/commit/bffe54d0aa9eaf96fc0f316d18f7411dd0e5c413)

Why do we still need it in the key then?

Reply:
I see, it looks like it's out of convenience so you can construct the key in one place, look at this property in another, and use a record struct to simplify adding the equality implementation, which requires having this property?

Reply:
Yep. Technically I could have extracted it to a separate parameter but this would require more refactoring.